### ISSUE TITLE
MultiDict.add should return this if no changes are made

### ISSUE THREAD

**Issue by bblfish** [scala/scala-collection-contrib#215]
The add method is defined as follows
```scala
  def add(key: K, value: V): MultiDict[K, V] =
    new MultiDict(elems.updatedWith(key) {
      case None     => Some(Set(value))
      case Some(vs) => Some(vs + value)
    })
```
but `elms.updatedWith` could easily return the same MultiDict if the value is already in the Set[V]. 
It would help if the signature made sure that the same object would be returned if nothing changed as that would allow one
to reduce unnecessary object creation (both in the call, and also in the creation of objects in the calling object. 
Otherwise one ends up creating a lot of garbage.

Something like this seems better:

```scala
def add(key: K, value: V): MultiDict[K, V] =
   val newMD = elems.updatedWith(key) {
      case None     => Some(Set(value))
      case Some(vs) => 
          val newVals = vs + value
          if newVals neq vs then Some(vs + value)
          else None
    }
    if newMD eq this.elems then this
    else new MultiDict(newMD)
```

**Comment by SethTisue** [comment]
@scala/collections ?

**Comment by bblfish** [comment]
yes, in scala collections for example [IntMap line 386](https://github.com/scala/scala/blob/4905801fc161300ba20bae4f7adf045b1698e831/src/library/scala/collection/immutable/IntMap.scala#L386), but I had found another example in a more widely used collection.

**Comment by bblfish** [comment]
I have a mini implementation of MultiDict for my needs in [org/w3/banana/quiver/util/MultiDict.scala](https://github.com/bblfish/banana-rdf/blob/milestoneW3/plantain/shared/src/main/scala/org/w3/banana/quiver/util/MultiDict.scala) with tests that pass and two implementations of Graph that use it whose tests also pass.
